app/browser/tabs.js

```diff
@@ -857,6 +887,86 @@ const api = {
     }
 
     return null
+  },
+
+  getNextActiveTab: (state, closeTabId) => {
+    if (!tabState.getByTabId(state, closeTabId)) {
+      return
+    }
+
+    const index = tabState.getIndex(state, closeTabId)
+    if (index === -1) {
+      return
+    }
+
+    const windowId = tabState.getWindowId(state, closeTabId)
+    if (windowId === windowState.WINDOW_ID_NONE) {
+      return
+    }
+
+    const lastActiveTabId = tabState.getTabsByLastActivated(state, windowId).last()
+    if (lastActiveTabId !== closeTabId && !tabState.isActive(state, closeTabId)) {
+      return
+    }
+
+    let nextTabId = tabState.TAB_ID_NONE
+    switch (getSetting(settings.TAB_CLOSE_ACTION)) {
+      case tabCloseAction.LAST_ACTIVE:
+        nextTabId = tabState.getLastActiveTabId(state, windowId)
+        break
+      case tabCloseAction.PARENT:
+        {
+          const openerTabId = tabState.getOpenerTabId(state, closeTabId)
+          if (openerTabId !== tabState.TAB_ID_NONE) {
+            nextTabId = openerTabId
+          }
+          break
+        }
+    }
+
+    // DEFAULT: always fall back to NEXT
+    if (nextTabId === tabState.TAB_ID_NONE) {
+      nextTabId = tabState.getNextTabIdByIndex(state, windowId, index)
+      if (nextTabId === tabState.TAB_ID_NONE) {
+        // no unpinned tabs so find the next pinned tab
+        nextTabId = tabState.getNextTabIdByIndex(state, windowId, index, true)
+      }
+    }
+
+    return nextTabId
+  },
+  debugTabs: (state) => {
+    console.log(tabState.getTabs(state)
+      .toJS()
+      .map((tab) => {
+        return {
+          tabId: tab.tabId,
+          index: tab.index,
+          windowId: tab.windowId,
+          active: tab.active
+        }
+      })
+      .sort((tab1, tab2) => {
+        if (tab1.windowId !== tab2.windowId) {
+          return tab1.windowId - tab2.windowId
+        }
+        if (tab1.index !== tab2.index) {
+          return tab1.index - tab2.index
+        }
+        return 0
+      }))
+  },
+  updateTabsStateForAttachedTab: (state, tabId) => {
+    const tabValue = getTabValue(tabId)
+    if (!tabValue) {
+      return
+    }
+    api.updateTabsStateForWindow(state, tabValue.get('windowId'))
+  },
+  updateTabsStateForWindow: (state, windowId) => {
+    tabState.getTabsByWindowId(state, windowId).forEach((tabValue) => {
+      updateTab(tabValue.get('tabId'))
```

Comment:
this is going to update async in another action. `updateTab` is designed to be called from an event, but since we're already in a reducer when this is called I think it would be better to update directly